src/crimson/common/interruptible_future.h

```diff
@@ -0,0 +1,740 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/future-util.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/when_all.hh>
+
+#include "crimson/common/errorator.h"
+
+namespace crimson::interruptible {
+
+struct ready_future_marker {};
+struct exception_future_marker {};
+
+template <typename InterruptCond>
+class interruptible_future_builder;
+
+template <typename InterruptCond>
+struct interruptor;
+
+template <typename InterruptCond>
+using InterruptCondRef = seastar::lw_shared_ptr<InterruptCond>;
+
+template <typename InterruptCond>
+InterruptCondRef<InterruptCond> interrupt_cond;
+
+template <typename InterruptCond, typename FutureType>
+class interruptible_future_detail {};
+
+template <typename FutureType>
+struct is_interruptible_future : public std::false_type {};
+
+template <typename InterruptCond, typename FutureType>
+struct is_interruptible_future<
+  interruptible_future_detail<
+    InterruptCond,
+    FutureType>>
+  : public std::true_type {};
+
+template <typename InterruptCond, typename Func, typename... T>
+auto call_with_interruption(
+  InterruptCondRef<InterruptCond> interrupt_condition,
+  Func&& func, T&&... args)
+{
+  using Result = std::invoke_result_t<Func, T...>;
+  using futurator_t = seastar::futurize<Result>;
+  if constexpr (sizeof...(T) == 1) {
+    if constexpr (InterruptCond::template is_interruption_v<
+	  std::tuple_element_t<0, std::tuple<T...>>>) {
+      return futurator_t::make_exception_future(
+	  std::get<0>(std::tuple(std::forward<T>(args)...)));
+    }
+    if constexpr (seastar::is_future<
+	  std::tuple_element_t<0, std::tuple<T...>>>::value) {
+      auto fut = std::get<0>(std::tuple(std::forward<T>(args)...));
+      std::exception_ptr eptr = fut.get_exception();
+      if (fut.failed() && interrupt_condition->is_interruption(eptr)) {
+	return futurator_t::make_exception_future(std::move(eptr));
+      }
+    }
+  }
+
+  if (interrupt_condition) {
+    auto [interrupt, fut] = interrupt_condition->template may_interrupt<
+      typename futurator_t::type>();
+    if (interrupt) {
+      return std::move(*fut);
+    }
+    interrupt_cond<InterruptCond> = interrupt_condition;
+  }
+
+  if constexpr (std::is_same_v<Result, void>) {
+    func(std::forward<T>(args)...);
+    return seastar::now();
+  } else {
+    auto fut = func(std::forward<T>(args)...);
+    if (interrupt_cond<InterruptCond>)
+      interrupt_cond<InterruptCond>.release();
```

Comment:
@athanatos "interrupt_cond" is cleared here:-)

Reply:
The seastar task that originally set interrupt_cond would have yielded the cpu long ago and many other tasks would have run, right?  This is within call_with_interruption which (for safe_then at least) gets called within the callback passed to core_type::safe_then.  In generally that would have happened after the wrapped future becomes available and not necessarily synchronously.

Reply:
Yes, but I don't think it matters. Because tasks captures the interrupt_cond when it is created which happens exactly within the same task that set the interrupt_cond, and when they get to run, they set the global interrupt_cond with their own captured one before executing the actual task body and clears interrupt_cond right before its finish. As seastar tasks can't be preempted, which means each task will reach its end before the cpu is yielded to other tasks once it gets to run, different continuation chains won't interfere with each other by means of interrupt conditions. Am I right?

Reply:
Sort of, but I'm worried that a safe_then call in a totally different task will pick up the interrupt_condition from a previous task.  If literally every one always calls enable_interrupt as soon as they regain control, it's ok, but auditing that in the current code base and *every future review* is to me at least clearly worse than simply annotating each safe_then call.

Reply:
> As *seastar tasks can't be preempted*, which means **each task will reach its end before the cpu is yielded to other tasks** once it gets to run, different continuation chains won't interfere with each other by means of interrupt conditions. Am I right?

I'm still reading through the `crimson::interruptible` and learning it but I think I will want to stress-test this assumption.
How about a situation when somebody runs `seastar::thread` and the flow reaches e.g. `seastar::future<...>::do_wait()`?